src/test/ui/consts/dangling-alloc-id-ice.rs

```diff
@@ -0,0 +1,15 @@
+// https://github.com/rust-lang/rust/issues/55223
+
+#![feature(const_let)]
+
+union Foo<'a> {
+    y: &'a (),
+    long_live_the_unit: &'static (),
+}
+
+const FOO: &() = { //~ ERROR this constant cannot be used
+    let y = ();
+    unsafe { Foo { y: &y }.long_live_the_unit }
```

Comment:
Same here, I don't understand why anything dangles. `y` is a local of the topmost frame, it should still be around when we intern? Or is there a `StorageDead` involved?

Reply:
It's not the topmost scope, there's a block inside the topmost scope, dropping `y` at the end of it. The MIR of

```rust
const FOO: () = {
    let y = ();
    y
};
```

is

```rust
const FOO: () ={
    let mut _0: ();                      // return place
    scope 1 {
    }
    scope 2 {
        let _1: ();                      // "y" in scope 2 at src/main.rs:4:9: 4:10
    }

    bb0: {                              
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:4:9: 4:10
        _0 = _1;                         // bb0[1]: scope 1 at src/main.rs:5:5: 5:6
        StorageDead(_1);                 // bb0[2]: scope 0 at src/main.rs:6:1: 6:2
        return;                          // bb0[3]: scope 0 at src/main.rs:3:1: 6:3
    }
}
```

Reply:
So it is a `StorageDead`.  I think we may want to change this in the future, but for now I agree this PR is better than the status quo.